lib/auth/grpcserver.go

```diff
@@ -1723,13 +1723,49 @@ func (g *GRPCServer) DeleteMFADevice(stream proto.AuthService_DeleteMFADeviceSer
 	// Find the device and delete it from backend.
 	devs, err := auth.GetMFADevices(ctx, user)
 	if err != nil {
-		return trace.Wrap(err)
+		return trail.ToGRPC(err)
+	}
+	authPref, err := auth.GetAuthPreference()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	var numTOTPDevs, numU2FDevs int
+	for _, d := range devs {
+		switch d.Device.(type) {
+		case *types.MFADevice_Totp:
+			numTOTPDevs++
+		case *types.MFADevice_U2F:
+			numU2FDevs++
+		default:
+			log.Warningf("Unknown MFA device type: %T", d.Device)
+		}
 	}
 	for _, d := range devs {
 		// Match device by name or ID.
 		if d.Metadata.Name != initReq.DeviceName && d.Id != initReq.DeviceName {
 			continue
 		}
+
+		// Make sure that the user won't be locked out by deleting the last MFA
+		// device. This only applies when the cluster requires MFA.
+		switch authPref.GetSecondFactor() {
+		case constants.SecondFactorOff, constants.SecondFactorOptional: // MFA is not required, allow deletion
+		case constants.SecondFactorOTP:
+			if numTOTPDevs == 1 {
+				return trail.ToGRPC(trace.BadParameter("cannot delete the last OTP device for this user; add a replacement device first to avoid getting locked out"))
+			}
```

Comment:
Does `SecondFactorOTP` mean there can be no MFA devices stored for the user? If a user had a single OTP device but he intended to delete some of the (now unusable) MFA devices, wouldn't this error be incorrectly returned since the type of the device to be deleted doesn't seem to be taken into account?

Analogically for `SecondFactorMFA` when deleting an OTP device.

Reply:
This block only returns an error if cluster requires `SecondFactorOTP` and there is exactly 1 OTP device registered (regardless of how many U2F devices there are).
Same thing for `SecondFactorU2F`, the error only happens when there's 1 U2F device, regardless of any OTP devices.

Maybe I misunderstand the question?

Reply:
Yes, I understand that.

My worry is that this error would be returned even if a user requested to delete an MFA device and not the last OTP device. In other words, the type of the device to be deleted should be checked and there should be no error when deleting an MFA device even if there is only a single OTP device left.

Reply:
OTP is a sub-type of MFA, not a separate group.
Each user has multiple MFA devices. Each of those devices is either U2F or OTP.
In the loop above I calculate the number of devices of each type (`numTOTPDevs` and `numU2FDevs`).
Does that clarify the logic?